File: engine/src/main/java/top/skyeyefast/mchjong/engine/TenhouReplay.java
package top.skyeyefast.mchjong.engine;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Tenhou /6 JSON interchange, mlog 2.3 (not compressed XML .mjlog).
 * Protocol reference: https://github.com/Equim-chan/tensoul/blob/main/convert.js
 * Serialization belongs to the application. Payments are recorded, never re-scored here.
 */
public final class TenhouReplay {
    private TenhouReplay() {}

    public static Map<String, Object> export(ReplayMatch match) {
        if (match.hands().isEmpty()) throw new IllegalArgumentException("There are no completed hands to export");
        var root = new LinkedHashMap<String, Object>();
        root.put("ver", "2.3");
        root.put("ref", match.id().toString());
        root.put("title", List.of("MCjhong · " + match.rules().name(), Instant.ofEpochMilli(match.startedAt()).toString()));
        var names = new ArrayList<String>();
        for (int i = 0; i < 4; i++) names.add(i < match.rules().players()
            ? match.participants().get(nativeSeat(match, i)).name() : "");
        root.put("name", names);
        root.put("ratingc", "PF" + match.rules().players());
        root.put("lobby", 0);
        root.put("dan", Collections.nCopies(4, ""));
        root.put("rate", Collections.nCopies(4, 0));
        root.put("sx", Collections.nCopies(4, ""));
        root.put("rule", Map.of("disp", (match.rules().sanma() ? "三" : "") + "南喰"
            + (match.redFives() == RedFives.NONE ? "" : "赤") + " · " + match.rules().name(),
            "aka51", match.rules().sanma() ? 0 : match.redFives().count(0),
            "aka52", match.redFives().count(1), "aka53", match.redFives().count(2)));
        root.put("log", match.hands().stream().map(hand -> hand(match, hand)).toList());
        if (match.complete()) {
            var last = match.hands().getLast();
            var score = new ArrayList<Number>();
            for (int i = 0; i < 4; i++) {
                if (i < match.rules().players()) {
                    int nativeSeat = nativeSeat(match, i);
                    score.add(last.finalSeats().get(nativeSeat).points());
                    score.add(last.finalScores().get(nativeSeat));
                } else { score.add(0); score.add(0); }
            }
            root.put("sc", score);
        }
        return root;
    }

    public static int tile(int physical) {
        int kind = Tile.kind(physical);
        return Tile.red(physical) ? 51 + kind / 9 : (kind / 9 + 1) * 10 + kind % 9 + 1;
    }

    private static int nativeSeat(ReplayMatch match, int external) {
        return (external + match.initialDealer()) % match.rules().players();
    }
    private static int externalSeat(ReplayMatch match, int nativeSeat) {
        return Math.floorMod(nativeSeat - match.initialDealer(), match.rules().players());
    }
    private static List<Integer> faces(List<Integer> physical) { return physical.stream().map(TenhouReplay::tile).toList(); }
    private static List<Integer> rotate(ReplayMatch match, List<Integer> values) {
        var result = new ArrayList<Integer>();
        for (int seat = 0; seat < 4; seat++) result.add(seat < match.rules().players() ? values.get(nativeSeat(match, seat)) : 0);
        return result;
    }

    private static List<Object> hand(ReplayMatch match, ReplayHand hand) {
        var result = new ArrayList<Object>();
        int players = match.rules().players();
        result.add(List.of(hand.round() / players * 4 + hand.round() % players, hand.honba(), hand.sticks()));
        result.add(rotate(match, hand.initialPoints()));
        result.add(faces(hand.dora()));
        result.add(faces(hand.ura()));
        var incoming = new ArrayList<List<Object>>();
        var outgoing = new ArrayList<List<Object>>();
        for (int seat = 0; seat < players; seat++) { incoming.add(new ArrayList<>()); outgoing.add(new ArrayList<>()); }
        for (var event : hand.events()) {
            if (event.kind() == ReplayHand.Kind.DORA || event.kind() == ReplayHand.Kind.RIICHI) continue;
            var draws = incoming.get(event.seat());
            var discards = outgoing.get(event.seat());
            switch (event.kind()) {
                case DRAW -> draws.add(tile(event.tile()));
                case DISCARD -> {
                    int symbol = event.tsumogiri() ? 60 : tile(event.tile());
                    discards.add(event.riichi() ? "r" + symbol : symbol);
                }
                case NUKI -> discards.add("f44");
                case MELD -> {
                    Meld meld = event.meld();
                    int who = externalSeat(match, event.seat());
                    int from = externalSeat(match, meld.fromSeat());
                    switch (meld.type()) {
                        case CHI, PON -> draws.add(call(meld.type(), who, from, meld.tiles(), meld.calledTile()));
                        case OPEN_KAN -> {
                            draws.add(call(meld.type(), who, from, meld.tiles(), meld.calledTile()));
                            discards.add(0);
                        }
                        case CLOSED_KAN -> {
                            var tiles = new ArrayList<>(faces(meld.tiles()));
                            Collections.sort(tiles);
                            discards.add("" + tiles.get(0) + tiles.get(1) + tiles.get(2) + "a" + tiles.get(3));
                        }
                        case ADDED_KAN -> {
                            var pon = new ArrayList<>(meld.tiles());
                            if (!pon.remove(Integer.valueOf(event.tile()))) throw new IllegalArgumentException("Missing added kan tile");
                            discards.add(call(Meld.Type.PON, who, from, pon, meld.calledTile())
                                .replace("p", "k" + tile(event.tile())));
                        }
                    }
                }
                default -> throw new IllegalStateException("Unexpected replay event");
            }
        }
        for (int seat = 0; seat < 4; seat++) {
            if (seat >= players) { result.add(List.of()); result.add(List.of()); result.add(List.of()); }
            else {
                int nativeSeat = nativeSeat(match, seat);
                result.add(faces(hand.initialHands().get(nativeSeat)));
                result.add(incoming.get(nativeSeat));
                result.add(outgoing.get(nativeSeat));
            }
        }
        result.add(result(match, hand));
        return result;
    }

    /** Marker position describes the source seat, and the following tile is the claimed tile. */
    static String call(Meld.Type type, int who, int from, List<Integer> tiles, int called) {
        var owned = new ArrayList<>(tiles);
        if (!owned.remove(Integer.valueOf(called))) throw new IllegalArgumentException("Missing called tile");
        owned.sort(Tile.ORDER);
        var tokens = new ArrayList<>(owned.stream().map(id -> Integer.toString(tile(id))).toList());
        int direction = Math.floorMod(who - from - 1, 4);
        if (type != Meld.Type.CHI && direction > 2) throw new IllegalArgumentException("A meld cannot call from itself");
        int index = type == Meld.Type.CHI ? 0 : type == Meld.Type.OPEN_KAN && direction == 2 ? 3 : direction;
        String marker = type == Meld.Type.CHI ? "c" : type == Meld.Type.PON ? "p" : "m";
        tokens.add(index, marker + tile(called));
        return String.join("", tokens);
    }

    private static List<Object> result(ReplayMatch match, ReplayHand hand) {
        var result = new ArrayList<Object>();
        if (hand.wins().isEmpty()) {
            result.add(switch (hand.result()) {
                case "exhaustive" -> "流局";
                case "nagashi" -> "流し満貫";
                case "nine_terminals" -> "九種九牌";
                case "four_winds" -> "四風連打";
                case "four_riichi" -> "四家立直";
                case "four_kans" -> "四開槓";
                case "triple_ron" -> "三家和";
                default -> throw new IllegalArgumentException("Unknown completed hand result: " + hand.result());
            });
            result.add(rotate(match, hand.deltas()));
            return result;
        }
        result.add("和了");
        for (var win : hand.wins()) {
            result.add(rotate(match, win.deltas()));
            var details = new ArrayList<Object>();
            int who = externalSeat(match, win.seat());
            details.add(who);
            details.add(win.from() < 0 ? who : externalSeat(match, win.from()));
            details.add(win.pao() < 0 ? who : externalSeat(match, win.pao()));
            details.add(score(win.score(), win.from() < 0, win.seat() == hand.dealer()));
            for (var yaku : win.yaku()) details.add(yakuName(yaku.name(), hand, win.seat(), match.rules().players())
                + "(" + (yaku.yakuman() ? "役満" : yaku.han() + "飜") + ")");
            if (win.dora() > 0) details.add("ドラ(" + win.dora() + "飜)");
            if (win.ura() > 0) details.add("裏ドラ(" + win.ura() + "飜)");
            if (win.redDora() > 0) details.add("赤ドラ(" + win.redDora() + "飜)");
            if (win.nukiDora() > 0) details.add("抜きドラ(" + win.nukiDora() + "飜)");
            result.add(details);
        }
        return result;
    }

    private static String score(HandScore score, boolean tsumo, boolean dealer) {
        int basic = score.ron() / (dealer ? 6 : 4);
        String prefix = basic >= 8000 ? score.yakuman() == 0 ? "数え役満"
            : score.yakuman() > 1 ? score.yakuman() + "倍役満" : "役満" : switch (basic) {
            case 6000 -> "三倍満";
            case 4000 -> "倍満";
            case 3000 -> "跳満";
            case 2000 -> "満貫";
            default -> score.fu() + "符" + score.han() + "飜";
        };
        return prefix + (!tsumo ? score.ron() + "点" : dealer ? score.tsumoChild() + "点∀"
            : score.tsumoChild() + "-" + score.tsumoDealer() + "点");
    }

    private static String yakuName(String name, ReplayHand hand, int seat, int players) {
        return switch (name) {
            case "SelfWind" -> "自風 " + "東南西北".charAt(Math.floorMod(seat - hand.dealer(), players));
            case "RoundWind" -> "場風 " + "東南西北".charAt(hand.round() / players);
            case "Tsumo" -> "門前清自摸和";
            case "Pinhu" -> "平和";
            case "Tanyao" -> "断幺九";
            case "Ipe" -> "一盃口";
            case "Haku" -> "役牌 白";
            case "Hatsu" -> "役牌 發";
            case "Chun" -> "役牌 中";
            case "Sanshoku" -> "三色同順";
            case "Ittsu" -> "一気通貫";
            case "Chanta" -> "混全帯幺九";
            case "Chitoi" -> "七対子";
            case "Toitoi" -> "対々和";
            case "Sananko" -> "三暗刻";
            case "Honroto" -> "混老頭";
            case "Sandoko" -> "三色同刻";
            case "Sankantsu" -> "三槓子";
            case "Shosangen" -> "小三元";
            case "Honitsu" -> "混一色";
            case "Junchan" -> "純全帯幺九";
            case "Ryanpe" -> "二盃口";
            case "Chinitsu" -> "清一色";
            case "Kokushi" -> "国士無双";
            case "Suanko" -> "四暗刻";
            case "Daisangen" -> "大三元";
            case "Tsuiso" -> "字一色";
            case "Shousushi" -> "小四喜";
            case "Lyuiso" -> "緑一色";
            case "Chinroto" -> "清老頭";
            case "Sukantsu" -> "四槓子";
            case "Churen" -> "九蓮宝燈";
            case "Daisushi" -> "大四喜";
            case "ChurenNineWaiting" -> "純正九蓮宝燈";
            case "SuankoTanki" -> "四暗刻単騎";
            case "KokushiThirteenWaiting" -> "国士無双十三面待ち";
            case "Richi" -> "立直";
            case "Ippatsu" -> "一発";
            case "Rinshan" -> "嶺上開花";
            case "Chankan" -> "槍槓";
            case "Haitei" -> "海底摸月";
            case "Houtei" -> "河底撈魚";
            case "WRichi" -> "ダブル立直";
            case "Tenhou" -> "天和";
            case "Chihou" -> "地和";
            case "Renhou" -> "人和";
            default -> throw new IllegalArgumentException("Unknown scoring yaku: " + name);
        };
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/TenhouReplay.kt
package top.skyeyefast.mchjong.engine

import java.time.Instant

/**
 * Tenhou /6 JSON interchange, mlog 2.3 (not compressed XML .mjlog).
 * Protocol reference: https://github.com/Equim-chan/tensoul/blob/main/convert.js
 * Serialization belongs to the application. Payments are recorded, never re-scored here.
 */
object TenhouReplay {
    @JvmStatic
    fun export(match: ReplayMatch): Map<String, Any> {
        require(match.hands.isNotEmpty()) { "There are no completed hands to export" }
        val root = linkedMapOf<String, Any>()
        root["ver"] = "2.3"
        root["ref"] = match.id.toString()
        root["title"] = listOf("MChjong · ${match.rules.name()}", Instant.ofEpochMilli(match.startedAt).toString())
        root["name"] = List(4) { seat ->
            if (seat < match.rules.players()) match.participants[nativeSeat(match, seat)].name else ""
        }
        root["ratingc"] = "PF${match.rules.players()}"
        root["lobby"] = 0
        root["dan"] = List(4) { "" }
        root["rate"] = List(4) { 0 }
        root["sx"] = List(4) { "" }
        root["rule"] = mapOf<String, Any>(
            "disp" to buildString {
                if (match.rules.sanma()) append("三")
                append(if (match.rules.matchLength() == 1) "東" else "南")
                if (match.rules.kuitan()) append("喰")
                if (match.redFives != RedFives.NONE) append("赤")
                append(" · ${match.rules.minHan()}飜縛り · ${match.rules.name()}")
            },
            "aka51" to if (match.rules.sanma()) 0 else match.redFives.count(0),
            "aka52" to match.redFives.count(1),
            "aka53" to match.redFives.count(2),
        )
        root["log"] = match.hands.map { hand(match, it) }
        if (match.complete) {
            val last = match.hands.last()
            val score = mutableListOf<Number>()
            for (seat in 0 until 4) {
                if (seat < match.rules.players()) {
                    val nativeSeat = nativeSeat(match, seat)
                    score.add(last.finalSeats[nativeSeat].points())
                    score.add(last.finalScores[nativeSeat])
                } else {
                    score.add(0)
                    score.add(0)
                }
            }
            root["sc"] = score
        }
        return root
    }

    @JvmStatic
    fun tile(physical: Int): Int {
        val kind = Tile.kind(physical)
        return if (Tile.red(physical)) 51 + kind / 9 else (kind / 9 + 1) * 10 + kind % 9 + 1
    }

    private fun nativeSeat(match: ReplayMatch, external: Int): Int =
        (external + match.initialDealer) % match.rules.players()

    private fun externalSeat(match: ReplayMatch, nativeSeat: Int): Int =
        Math.floorMod(nativeSeat - match.initialDealer, match.rules.players())

    private fun faces(physical: List<Int>): List<Int> = physical.map(::tile)

    private fun rotate(match: ReplayMatch, values: List<Int>): List<Int> =
        List(4) { seat -> if (seat < match.rules.players()) values[nativeSeat(match, seat)] else 0 }

    private fun hand(match: ReplayMatch, hand: ReplayHand): List<Any> {
        val result = mutableListOf<Any>()
        val players = match.rules.players()
        result.add(listOf(hand.round / players * 4 + hand.round % players, hand.honba, hand.sticks))
        result.add(rotate(match, hand.initialPoints))
        result.add(faces(hand.dora))
        result.add(faces(hand.ura))
        val incoming = MutableList(players) { mutableListOf<Any>() }
        val outgoing = MutableList(players) { mutableListOf<Any>() }
        for (event in hand.events) {
            if (event.kind == ReplayHand.Kind.DORA || event.kind == ReplayHand.Kind.RIICHI) continue
            val draws = incoming[event.seat]
            val discards = outgoing[event.seat]
            when (event.kind) {
                ReplayHand.Kind.DRAW -> draws.add(tile(event.tile))
                ReplayHand.Kind.DISCARD -> {
                    val symbol = if (event.tsumogiri) 60 else tile(event.tile)
                    discards.add(if (event.riichi) "r$symbol" else symbol)
                }
                ReplayHand.Kind.NUKI -> discards.add("f44")
                ReplayHand.Kind.MELD -> {
                    val meld = event.meld!!
                    val who = externalSeat(match, event.seat)
                    val from = externalSeat(match, meld.fromSeat())
                    when (meld.type()) {
                        Meld.Type.CHI, Meld.Type.PON -> draws.add(call(meld.type(), who, from, meld.tiles(), meld.calledTile()))
                        Meld.Type.OPEN_KAN -> {
                            draws.add(call(meld.type(), who, from, meld.tiles(), meld.calledTile()))
                            discards.add(0)
                        }
                        Meld.Type.CLOSED_KAN -> {
                            val tiles = faces(meld.tiles()).sorted()
                            discards.add("${tiles[0]}${tiles[1]}${tiles[2]}a${tiles[3]}")
                        }
                        Meld.Type.ADDED_KAN -> {
                            val pon = meld.tiles().toMutableList()
                            require(pon.remove(event.tile)) { "Missing added kan tile" }
                            discards.add(
                                call(Meld.Type.PON, who, from, pon, meld.calledTile())
                                    .replace("p", "k${tile(event.tile)}"),
                            )
                        }
                    }
                }
                ReplayHand.Kind.DORA, ReplayHand.Kind.RIICHI -> throw IllegalStateException("Unexpected replay event")
            }
        }
        for (seat in 0 until 4) {
            if (seat >= players) {
                repeat(3) { result.add(emptyList<Any>()) }
            } else {
                val nativeSeat = nativeSeat(match, seat)
                result.add(faces(hand.initialHands[nativeSeat]))
                result.add(incoming[nativeSeat])
                result.add(outgoing[nativeSeat])
            }
        }
        result.add(result(match, hand))
        return result
    }

    /** Marker position describes the source seat, and the following tile is the claimed tile. */
    @JvmStatic
    fun call(type: Meld.Type, who: Int, from: Int, tiles: List<Int>, called: Int): String {
        val owned = tiles.toMutableList()
        require(owned.remove(called)) { "Missing called tile" }
        owned.sortWith(Tile.ORDER)
        val tokens = owned.map { tile(it).toString() }.toMutableList()
        val direction = Math.floorMod(who - from - 1, 4)
        require(type == Meld.Type.CHI || direction <= 2) { "A meld cannot call from itself" }
        val index = when {
            type == Meld.Type.CHI -> 0
            type == Meld.Type.OPEN_KAN && direction == 2 -> 3
            else -> direction
        }
        val marker = when (type) {
            Meld.Type.CHI -> "c"
            Meld.Type.PON -> "p"
            else -> "m"
        }
        tokens.add(index, marker + tile(called))
        return tokens.joinToString("")
    }

    private fun result(match: ReplayMatch, hand: ReplayHand): List<Any> {
        val result = mutableListOf<Any>()
        if (hand.wins.isEmpty()) {
            result.add(
                when (hand.result) {
                    "exhaustive" -> "流局"
                    "nagashi" -> "流し満貫"
                    "nine_terminals" -> "九種九牌"
                    "four_winds" -> "四風連打"
                    "four_riichi" -> "四家立直"
                    "four_kans" -> "四開槓"
                    "triple_ron" -> "三家和"
                    else -> throw IllegalArgumentException("Unknown completed hand result: ${hand.result}")
                },
            )
            result.add(rotate(match, hand.deltas))
            return result
        }
        result.add("和了")
        for (win in hand.wins) {
            result.add(rotate(match, win.deltas))
            val details = mutableListOf<Any>()
            val who = externalSeat(match, win.seat)
            details.add(who)
            details.add(if (win.from < 0) who else externalSeat(match, win.from))
            details.add(if (win.pao < 0) who else externalSeat(match, win.pao))
            details.add(score(win.score, win.from < 0, win.seat == hand.dealer))
            for (yaku in win.yaku) {
                val value = if (yaku.yakuman) "役満" else "${yaku.han}飜"
                details.add("${yakuName(yaku.name, hand, win.seat, match.rules.players())}($value)")
            }
            if (win.dora > 0) details.add("ドラ(${win.dora}飜)")
            if (win.ura > 0) details.add("裏ドラ(${win.ura}飜)")
            if (win.redDora > 0) details.add("赤ドラ(${win.redDora}飜)")
            if (win.nukiDora > 0) details.add("抜きドラ(${win.nukiDora}飜)")
            result.add(details)
        }
        return result
    }

    private fun score(score: HandScore, tsumo: Boolean, dealer: Boolean): String {
        val basic = score.ron() / if (dealer) 6 else 4
        val prefix = if (basic >= 8000) {
            when {
                score.yakuman() == 0 -> "数え役満"
                score.yakuman() > 1 -> "${score.yakuman()}倍役満"
                else -> "役満"
            }
        } else {
            when (basic) {
                6000 -> "三倍満"
                4000 -> "倍満"
                3000 -> "跳満"
                2000 -> "満貫"
                else -> "${score.fu()}符${score.han()}飜"
            }
        }
        return prefix + if (!tsumo) {
            "${score.ron()}点"
        } else if (dealer) {
            "${score.tsumoChild()}点∀"
        } else {
            "${score.tsumoChild()}-${score.tsumoDealer()}点"
        }
    }

    private fun yakuName(name: String, hand: ReplayHand, seat: Int, players: Int): String = when (name) {
        "SelfWind" -> "自風 " + "東南西北"[Math.floorMod(seat - hand.dealer, players)]
        "RoundWind" -> "場風 " + "東南西北"[hand.round / players]
        "Tsumo" -> "門前清自摸和"
        "Pinhu" -> "平和"
        "Tanyao" -> "断幺九"
        "Ipe" -> "一盃口"
        "Haku" -> "役牌 白"
        "Hatsu" -> "役牌 發"
        "Chun" -> "役牌 中"
        "Sanshoku" -> "三色同順"
        "Ittsu" -> "一気通貫"
        "Chanta" -> "混全帯幺九"
        "Chitoi" -> "七対子"
        "Toitoi" -> "対々和"
        "Sananko" -> "三暗刻"
        "Honroto" -> "混老頭"
        "Sandoko" -> "三色同刻"
        "Sankantsu" -> "三槓子"
        "Shosangen" -> "小三元"
        "Honitsu" -> "混一色"
        "Junchan" -> "純全帯幺九"
        "Ryanpe" -> "二盃口"
        "Chinitsu" -> "清一色"
        "Kokushi" -> "国士無双"
        "Suanko" -> "四暗刻"
        "Daisangen" -> "大三元"
        "Tsuiso" -> "字一色"
        "Shousushi" -> "小四喜"
        "Lyuiso" -> "緑一色"
        "Chinroto" -> "清老頭"
        "Sukantsu" -> "四槓子"
        "Churen" -> "九蓮宝燈"
        "Daisushi" -> "大四喜"
        "ChurenNineWaiting" -> "純正九蓮宝燈"
        "SuankoTanki" -> "四暗刻単騎"
        "KokushiThirteenWaiting" -> "国士無双十三面待ち"
        "Richi" -> "立直"
        "Ippatsu" -> "一発"
        "Rinshan" -> "嶺上開花"
        "Chankan" -> "槍槓"
        "Haitei" -> "海底摸月"
        "Houtei" -> "河底撈魚"
        "WRichi" -> "ダブル立直"
        "Tenhou" -> "天和"
        "Chihou" -> "地和"
        "Renhou" -> "人和"
        else -> throw IllegalArgumentException("Unknown scoring yaku: $name")
    }
}
